File: C_Sharp_code/Example/WindowsFormsApp/CS_demo21/Lion.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace CS_demo21
{
    public class Lion : Animal, ICarnivore, IComparable
    {
        private bool hungry;
        public string name;
        public Lion()
        {
            hungry = true;
        }
        public Lion(string name)
        {
            this.name = name;
        }

        bool ICarnivore.IsHungry
        {
            get
            {
                return hungry;
            }
        }

        public override void sleep()
        {
            Console.WriteLine("獅子在睡覺");
        }

        void ICarnivore.Eat(Animal victim)
        {
            Console.WriteLine("獅子吃獵物 " + victim);
        }

        Animal ICarnivore.Hunt()
        {
            //狩獵到羚羊
            return new Antelope();
        }

        public int CompareTo(object obj)
        {
            Lion x;
            x = (Lion)obj;
            //Lion自己本身的物件name和傳遞進來的物件name做比較
            //本身大於傳遞進來,回傳正值,反之則回傳負值
            return this.name.CompareTo(x.name);
        }
    }
}

File: C_Sharp_code/Example/WindowsFormsApp/VB_demo22/Lion.vb
Public Class Lion
    Inherits Animal
    Implements ICarnivore, IComparable

    Private isHungry As Boolean
    Public name As String

    Private ReadOnly Property ICarnivore_IsHungry As Boolean Implements ICarnivore.IsHungry
        Get
            Return isHungry
        End Get
    End Property

    Public Sub New()
        isHungry = True
    End Sub
    Public Sub New(ByVal argName As String)
        Me.name = argName
    End Sub

    Public Sub Eat(Victim As Animal) Implements ICarnivore.Eat
        Console.WriteLine("獅子吃獵物 " & Victim.ToString())
    End Sub

    Public Function Hunt() As Animal Implements ICarnivore.Hunt
        Return New Antelope()
    End Function

    Public Overrides Sub Sleep()
        Console.WriteLine("獅子在睡覺")
    End Sub

    Public Function CompareTo(obj As Object) As Integer Implements IComparable.CompareTo
        Dim x As Lion
        'Lion自己本身的物件name和傳遞進來的物件name做比較
        '本身大於傳遞進來,回傳正值,反之則回傳負值
        x = CType(obj, Lion)
        Return Me.name.CompareTo(x.name) * -1
    End Function
End Class
